### ISSUE TITLE
The vsphere metric `vsphere_vm_sys_uptime_latest` `guest` label value doesn't reflect the actual guest operating system

### ISSUE THREAD

**Issue by imperatives** [influxdata/telegraf#14616]
### Use Case

The telegraf vsphere plugin appears to set the the guest os version to Config.GuestId, which is the value that is set when creating the VM and can differ from the actual guest os version.  In order to more accurately reflect the state of the guest Guest.GuestId should be used since this is derived from vmtools.

### Expected behavior

When the current os version is `rhel8_64` the `vsphere_vm_sys_uptime_latest{guesthostname="<guesthostname>",guest="rhel7_64"}` metric `guest` label value should be `rhel8_64`.  Since the guest was originally created with `rhel7_64` the value is incorrectly reported.

### Actual behavior

The `vsphere_vm_sys_uptime_latest{guesthostname="<guesthostname>",guest="rhel7_64"}` metric `guest` label is incorrect since it uses `Config.GuestId`.

### Additional info

Example: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/vsphere/finder.go#L260 `config.guestId` should be changed to `guest.guestId`.  Ideally the default would be to use `guest.guestId` and fallback to `config.guestId` if it isn't defined.  This would prevent failures when vmtools isn't installed on the guest or it isn't reporting properly.
Here is an example: https://github.com/kubevirt/vm-import-operator/blob/master/pkg/providers/vmware/os/os-finder.go#L28-L55

**Comment by powersj** [comment]
@imperatives,

Do you have a way of testing both of the following cases: 1) where the `guest.guestId` value is different than `config.guestId` and 2) where the `guest.guestId` value is not present?

I think if we collect the `r.Guest.GuestId`, we could check if it exists [here](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/vsphere/endpoint.go#L841-L844) and use it.

This could be potentially disruptive of current users as it would involve changing tags, so we would also need to consider putting this behind a config option.

**Comment by imperatives** [comment]
Thanks @powersj Yes, I have the ability to test both cases.  I used the following powershell command to identify the differences for a vm in our environment.  The vm was running RHEL8, but reporting as RHEL7 in the metrics. ```Get-VM -name <guestname> | Get-View -Property @("Name", "Config.GuestId", "Guest.GuestId") | Select -Property Name, @{N="Configured OS";E={$_.Config.GuestId}}, @{N="Running OS";E={$_.Guest.GuestId}} | Format-Table -AutoSize```.  For the second case I believe we can simply disable/remove the Open VM Tools from the vm so that it can't report back to vcenter.  Then we can confirm the results for the `config.guestId`.  I'll post an update on Monday.

**Comment by powersj** [comment]
@imperatives,

I put up https://github.com/influxdata/telegraf/pull/14635 with the ability to grab the newer value if it is set. Can you give it a try? There will be artifacts attached to the PR in 20-30mins after this post.

Thanks!

**Comment by imperatives** [comment]
@powersj 
Great!  I was able to deploy that artifact to our test environment and here are my findings:
1. Using powershell I confirmed the host in question was reporting different values for Config.GuestId (`rhel7_64`) and Guest.GuestId (`rhel8_64`)
2. Using Telegraf `telegraf` version `1.28.5` the metric `vsphere_vm_sys_uptime_latest{guesthostname="<hostname>"}` produced the guest label for the host as `guest="rhel7_64"`.
3.  When using the attached telegraf artifact the metric `vsphere_vm_sys_uptime_latest{guesthostname="<hostname>"}` produced the guest label for the host as `guest="rhel8_64"` (success).
4. Then I removed the Open VM Tools from the host and confirmed via powershell that the `Guest.GuestId` was no longer present.
5. When I queried our time-series database (Grafana Mimir) no new metrics were populated for that particular host.  I stopped and started the telegraf agent to confirm as well without any change.  
6. Finally I reinstalled the Open VM Tools and metrics started to populate one again.

It appears that these changes may break metrics for hosts that don't have vm tools installed (i.e., guest.guestid is absent).

**Comment by imperatives** [comment]
@powersj 
When no vm tools are installed the guest.guestid query doesn't return anything from vsphere.  Only the config.guestId returns a value.
```
Get-VM -name <guestname>| Get-View -Property @("Name", "Config.GuestId", "Guest.GuestId") | Select -Property Name, @{N="Configured OS";E={$_.Config.GuestId}}, @{N="Running OS";E={$_.Guest.GuestId}} | Format-Table -AutoSize

Name         Configured OS Running OS
----         ------------- ----------
<hostname> rhel7_64Guest 
```

I ran  `./telegraf --test` and it clued me into the `source` label.  As you can see the `guesthostname` is absent from the measurement, but the vm name is in the `source` label. 
```
vsphere_vm_cpu,clustername=dev,cpu=1,dcname=dev,esxhostname=<esxhostname>,guest=rhel7_64,host=<host>,rpname=<rpname>,source=<guestname>,vcenter=<vcenter_ip>,vmname=<vmname> costop_summation=0i,readiness_average=0.02,ready_summation=5i,usagemhz_average=37i,used_summation=374i 1706746680000000000
```
When running from either the existing telegraf version or the attached artifact no metrics would appear for the test host under the label `guesthostname`.  However, it was populating metrics for the host on the label `source`.  In my previous message I was only filtering on `guesthostname`, which is why I didn't see any metrics for the test host.  I don't know if this is expected behavior.

To summarize, the `guest.guestid` change you provided appears to be working as expected.  I just wasn't aware that the `guesthostname` would disappear from the measurement if vm tools installed on the vm.  Since this behavior is present in the existing version I don't see that as related to this issue.  Let me know if I can provide any additional info.  Thanks again for putting this fix together so quickly!
